File: Leetcode Challenge/december2k22/Binary Tree Maximum Path Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int max_sum=INT_MIN;
    int dfs(TreeNode* root)
    {
        if(!root)return 0;
        int l=max(dfs(root->left),0);
        int r=max(dfs(root->right),0);
        int new_price=root->val+l+r;
        max_sum=max(max_sum,new_price);
        return root->val+max(l,r);
    }
    int maxPathSum(TreeNode* root) {
        dfs(root);
        return max_sum;
    }
};

File: Leetcode Challenge/december2k22/Python/Binary Tree Maximum Path Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.maxi=-99999

        def maxpath(root):

            if root is None:
                return 0


            lh=max(0,maxpath(root.left))
            rh=max(0,maxpath(root.right))

            self.maxi=max(self.maxi,root.val+lh+rh)




            

            return root.val+max(lh,rh)




        maxpath(root)
        return self.maxi
